### PR TITLE
feat: add keymap size (@butterflycup)

### PR DESCRIPTION
**Pull request by butterflycup**
The keymap cannot be resized, it can sometimes feel pretty small, especially when learning a new layout. This adds a setting called "Keymap Size" in which you can change the size of the keymap (1 being default, 0.5 minimum and 3.5 as maximum as anything higher wouldn't fit anyways)

### CODE REVIEW

File: frontend/.vscode/settings.json
Comment (Miodec) [inline frontend/.vscode/settings.json:1]: Remove this file please.

File: frontend/src/html/pages/settings.html [hunk 1]
```diff
@@ -1051,6 +1051,21 @@
     </div>
     <div class="sectionSpacer"></div>
   </div>
+  <div class="section" data-config-name="keymapSize">
+    <div class="groupTitle">
+      <i class="fas fa-keyboard"></i>
+      <span>keymap size</span>
+    </div>
+    <div class="text">Change the scale of the keymap.</div>
```
Comment (Miodec) [inline frontend/src/html/pages/settings.html:1059]: scale > size

File: frontend/src/ts/config.ts [hunk 2]
```diff
@@ -1583,6 +1583,31 @@ export function setKeymapShowTopRow(
...
 }
 
+export function setKeymapSize(keymapSize: number, nosave?: boolean): boolean {
+  if (!isConfigValueValid("keymap size", keymapSize, ["number"])) {
+    return false;
+  }
+
+  if (keymapSize < 0) {
+    keymapSize = 1;
+  }
```
Comment (Miodec) [inline frontend/src/ts/config.ts:1593]: This allows the size to be 0, which could be confusing if the keymap is enabled but still invisible. I would limit this to 0.5

File: frontend/src/html/pages/settings.html [hunk 3]
```diff
@@ -1051,6 +1051,21 @@
...
+    <div class="inputs">
+      <div class="inputAndButton">
+        <input type="number" placeholder="keymap size" class="input" min="0" />
+        <button class="save no-auto-handle">
```
Comment (Miodec) [inline frontend/src/html/pages/settings.html:1063]: Update min property to 0.5 and add max with 3.5

**Review by Miodec — CHANGES_REQUESTED** [review]

**Comment by Miodec** [comment]
By scale > size i meant to change the text "scale" to "size", not changing everything to scale :D Sorry for the ambigous comment.

**Comment by Miodec** [comment]
image

**Comment by butterflycup** [comment]
My bad, though in the end I guess scale makes more sense than size anyway, which one do you prefer ?

**Comment by Miodec** [comment]
Other settings use size - lets stick with that.

**Review by Miodec — CHANGES_REQUESTED** [review]
Please hide the settings group if keymap is not enabled (just like the other settings which alter the keymap)

Also, what do you think about changing this to a slider (just like the custom background filter settings?)

image

image

**Comment by butterflycup** [comment]
Keymap setting groups now hides, I prefer number input instead of slider. I'd go for slider if the changes were visible in real time (like for the custom background), but since the keymap is not visible in the settings page (and there's no slider in command line) I prefer number input, though I think both works. Maybe it'd be good to ask other people's opinion?

**Comment by Miodec** [comment]
> Keymap setting groups now hides, I prefer number input instead of slider. I'd go for slider if the changes were visible in real time (like for the custom background), but since the keymap is not visible in the settings page (and there's no slider in command line) I prefer number input, though I think both works. Maybe it'd be good to ask other people's opinion?

Well, there is no slider for any of the settings in the commandline.

The slider makes more sense in this case because the value of this setting is limited between two values, and the range is quite small. @fehmer also liked the slider a bit more.

**Comment by butterflycup** [comment]
I am way out of my comfort zone and I'm guessing this is why there is conflict with my fork (I couldn't figure out why), I'd appreciate the help a lot

**Comment by fehmer** [comment]
> I am way out of my comfort zone and I'm guessing this is why there is conflict with my fork (I couldn't figure out why), I'd appreciate the help a lot

hi @butterflycup , i'm done with syncing your branch with monkeytype master and adapted to the new approach. You can take it from here.

**Comment by Miodec** [comment]
> I think waiting for a slider implementation would unnecessarily postpone this feature of keymap resizing.

Converting to a slider would take like half an hour max.

**Comment by Miodec** [comment]
Ill go ahead and convert to a slider

**Comment by butterflycup** [comment]
Thank you all, I did the best I could, and while Miodec is right that it would take "half an hour max", I have to admit I am a noob and it wouldn't be as easy for me. Thank you for taking over

**Comment by Miodec** [comment]
> Thank you all, I did the best I could, and while Miodec is right that it would take "half an hour max", I have to admit I am a noob and it wouldn't be as easy for me. Thank you for taking over

No worries, it was a big ask to be fair.

**Pull request merged**